python/objToJSON.c

```diff
@@ -746,6 +746,30 @@ static char *Object_iterGetName(JSOBJ obj, JSONTypeContext *tc, size_t *outLen)
   return GET_TC(tc)->iterGetName(obj, tc, outLen);
 }
 
+
+static const char *_PyUnicodeToChars(PyObject *obj, size_t *_outLen)
+{
+  // helper for indent only
+  PyObject *newObj;
+#ifndef Py_LIMITED_API
+  if (PyUnicode_IS_COMPACT_ASCII(obj))
+  {
+    Py_ssize_t len;
+    const char *data = PyUnicode_AsUTF8AndSize(obj, &len);
+    *_outLen = len;
+    return data;
+  }
+#endif
+  newObj = PyUnicode_AsUTF8String(obj);
+  if(!newObj)
+  {
+    return NULL;
+  }
+
+  *_outLen = PyBytes_Size(newObj);
+  return PyBytes_AsString(newObj);
```

Comment:
As you mentioned, `PyUnicode_AsUTF8AndSize` and `PyBytes_AsString` don't require dereferencing. But as far as I can tell, `PyUnicode_AsUTF8String` does. So there'd need to be a `Py_DECREF(newObj)` call between `PyBytes_AsString` and returning, I think.

Reply:
I'll look more into that. Also wrt to this code, is there any better way of implementing the `_PyUnicodeToChars` logic? I took this logic from elsewhere in the file that was doing a similar, but not identical thing. I'm curious what best practices are here.

Reply:
It seems like this is indeed the way to do it. Looks like the other similar code keeps a reference to `newObj` and then later calls `Py_XDECREF` on it, by the way.

Also, sidenote: `PyUnicode_AsUTF8AndSize` is part of the stable ABI since Python 3.10, so once that's the minimum supported version, only the code in the `#ifndef` will be needed.

Reply:
I just looked at this again as I'm trying to fix the issues with surrogates. `PyBytes_AsString` returns a pointer to the internal `char*` buffer. This means the `PyObject` can't be `DECREF`d until the buffer is no longer needed. This is why the similar existing function stores the `newObj` in the context struct.

It should be possible to refactor `PyUnicodeToUTF8` slightly to make it reusable for this (and it'll be useful for other things as well, e.g. `separators`). Then the `PyObject` would be kept and `DECREF`d at the end of `objToJSON`. I'll incorporate that refactor in my surrogate fix since it requires changes to that function anyway.

Reply:
→ #530

Reply:
The fix for the surrogates unblocked me in terms of what I could do in this PR. Using `PyUnicode_AsEncodedString` was the key. I rebased on your branch so I could get those changes, and the tests are now at least working locally. Let's see how they do on the CI. This PR probably need a bit of cleanup because I did a bunch of print debugging.

Reply:
You can use my `PyUnicodeToUTF8Raw`, no need for the duplicate function. You will need to take care of `DECREF`ing the `PyBytes` object at the end though as mentioned above (also in case of `objToJSON` returning early!).